include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,454 @@
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+				 const struct tm *timeptr);
+
+static inline int z_impl_rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+					    const struct tm *timeptr)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_set_time(dev, id, mask, timeptr);
+}
+
+/**
+ * @brief API for getting RTC alarm time
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm
+ * @param mask Destination for mask of fields which are enabled in the alarm time
+ * @param timeptr Destination for the alarm time
+ *
+ * @note bits in mask are stored in RTC_ALARM_TIME_MASK defines
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				 struct tm *timeptr);
+
+static inline int z_impl_rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+					    struct tm *timeptr)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_get_time(dev, id, mask, timeptr);
+}
+
+/**
+ * @brief API for testing if RTC alarm is pending
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm to test
+ *
+ * @return 1 if alarm is pending
+ * @return 0 if alarm is not pending
+ * @return -EINVAL if id is out of range
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_is_pending(const struct device *dev, uint16_t id);
+
+static inline int z_impl_rtc_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_is_pending(dev, id);
+}
+
+/**
+ * @brief API for clearing RTC alarm pending status
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm for which the pending status will be cleared
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_clear_pending(const struct device *dev, uint16_t id);
+
+static inline int z_impl_rtc_alarm_clear_pending(const struct device *dev, uint16_t id)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_clear_pending(dev, id);
+}
+
+/**
+ * @brief API for setting alarm IRQ callback
+ *
+ * @param dev Device instance
+ * @param callback Callback called when alarm IRQ occurs
+ * @param user_data Optional user data passed to callback
```

Comment:
1. No, when alarm IRQ is called, the pending state is cleared.
2. That happens automatically after callback
3. Yes, the callback is the interrupt, so it is no longer pending when callback has returned.
4. No, work generally should not be performed in an IRQ handler, that work should be delegated using the sys work queue for example.

Reply:
This function will be called from irq handler context?  This does not work for RTCs that are external devices, since it's necessary to query the device over I2C or SPI and that requires a work queue.

If the pending status is clear after the handler, there is a race where after the handler a new alarm is generated, which is the then cleared without being handled.

Reply:
My thinking is that an external RTC would have an output connected to a GPIO, and that GPIO would trigger the interrupt, not that one would be polling the bit in the RTCs register. If the bit is polled, then I would expect the application to have to compromise and manually poll with whatever frequency it found appropriate.